feat(npm): support constraintsFiltering=strict (#22447)
renovatebot/renovate: docs/usage/configuration-options.md

- `none`: No release filtering (all releases allowed)
- `strict`: If the release's constraints match the package file constraints, then it's included

More advanced filtering options may come in future.

> Note: There must be a `constraints` object in your Renovate config, or constraints detected from package files, for this to work.

Reviewer: ```suggestion
<!-- prettier-ignore -->
!!! note
    There must be a `constraints` object in your Renovate config, or constraints detected from package files, for this to work.

```

Choose your adventure:

1. Make a proper `note` admonition -> apply suggestion
2. Drop the `Note:` part of the sentence, and rely on the `There must be ...` part to explain the requirement. -> edit suggestion and then apply
3. Do nothing.

> Note: There must be a `constraints` object in your Renovate config, or constraints detected from package files, for this to work.

Author: ```suggestion
There must be a `constraints` object in your Renovate config, or constraints detected from package files, for this to work.
```

This feature is limited to `packagist`, `npm`, and `pypi` datasources.

<!-- prettier-ignore -->
!!! warning
    Enabling this feature may result in many package updates being filtered out silently.
    See below for a description of how it works.

When `constraintsFiltering=strict`, the following logic applies:

- Are there `constraints` for this repository, either detected from source or from config?
- Does this package's release declare constraints of its own (e.g. `engines` in Node.js)?
- If so, filter out this release unless the repository constraint is a _subset_ of the release constraint

Here are some examples:

> | project                  | dependency   | result   |

Reviewer: ```suggestion
| Project                  | Dependency   | Result   |
```

I like the table! 🥳 Let's capitalize the headers though.

What do you mean with `Project` and `Dependency`? Do you mean something like `Detected constraint` and `Dependency constraint`?

After fixing the table headers you should run `yarn prettier-fix` to format the table. 😉

|--------------------------|--------------|----------|
| `18`                     | `16 \|\| 18` | allowed  |
| `^18.10.0`               | `>=18`       | allowed  |
| `^16.10.0 \|\| >=18.0.0` | `>= 16.0.0`  | allowed  |
| `>=16`                   | `16 \|\| 18` | filtered |
| `16`                     | `^16.10.0`   | filtered |

When using with `npm`, we recommend you:

> - Use with `dependencies`, not `devDependencies` (usually you do not need to be strict about development dependencies)

Reviewer: ```suggestion
- Use `constraintsFiltering` on `dependencies`, not `devDependencies` (usually you do not need to be strict about development dependencies)
```

`Use with dependencies` is fuzzy for me, I think we mean what I'm suggesting above?

> - Do not enable `rollbackPrs` at the same time (otherwise your _current_ version may be rolled back if it's incompatible)

Reviewer: ```suggestion
- Do _not_ enable `rollbackPrs` at the same time (otherwise your _current_ version may be rolled back if it's incompatible)
```

Let's emphasize the _not_ part.

> - Do _not_ enable `rollbackPrs` at the same time (otherwise your _current_ version may be rolled back if it's incompatible)

Reviewer: ```suggestion
- Set `rollbackPrs` to `false` for the `npm` manager with a `packageRules` config, to prevent your _current_ version from getting rolled back if it's incompatible
```

Let's say what the user must do, instead of what they must avoid. Might be good to add this config example as well:

```json
{
  "packageRules": [
    {
      "matchManagers": ["npm"],
      "rollbackPrs": false
    }
  ]
}
```

Author: Starting in v36, rollbackPrs will be false by default for npm, so at that point it is correct IMO to say "don't enable it" instead of "disable it"

Reviewer: This PR targets `main` though, where `rollbackPrs` are still enabled for `npm`... 😉 

Options:

1. Point this PR at the `v36` branch, so your current docs are correct, or
2. Keep targeting `main`, apply my suggestion, merge PR, and create a PR targeting `v36` to change the text again 😄

- `none`: No release filtering (all releases allowed)
- `strict`: If the release's constraints match the package file constraints, then it's included

More advanced filtering options may come in future.

Note: There must be a `constraints` object in your Renovate config, or constraints detected from package files, for this to work.

> This feature is limited to `packagist`, `npm`, and `pypi` datasources.

Author: FYI: The statement including `npm` above was wrong until now

<!-- prettier-ignore -->
!!! warning
    Enabling this feature may result in many package updates being filtered out silently.
    See below for a description of how it works.

When `constraintsFiltering=strict`, the following logic applies:

- Are there `constraints` for this repository, either detected from source or from config?
- Does this package's release declare constraints of its own (e.g. `engines` in Node.js)?
- If so, filter out this release unless the repository constraint is a _subset_ of the release constraint

Here's some examples of when a release is allowed:

- The `package.json` declares its `engines.node` as `18` which is a subset of the package release `16 || 18`
- The `package.json` declares its `engines.node` as `^18.10.0` which is a subset of the package release `>=18`
- The `package.json` declares its `engines.node` as `^16.10.0 || >=18.0.0` which is a subset of the package release `>= 16.0.0`

Here's some examples of when a release is filtered out (disallowed):

- The `package.json` declares its `engines.node` as `>=16` while the package release has a narrower `16 || 18`

> - The `package.json` declares its `engines.node` as `16` while the package release has a narrower `^16.10.0`

Reviewer: Can we try a table layout instead of a list?

Reviewer: This comment needs to be resolved by a maintainer, I don't have rights to resolve review comments.

> Here's some examples of when a release is allowed:
> 
> - The `package.json` declares its `engines.node` as `18` which is a subset of the package release `16 || 18`
> - The `package.json` declares its `engines.node` as `^18.10.0` which is a subset of the package release `>=18`
> - The `package.json` declares its `engines.node` as `^16.10.0 || >=18.0.0` which is a subset of the package release `>= 16.0.0`
> 
> Here's some examples of when a release is filtered out (disallowed):
> 
> - The `package.json` declares its `engines.node` as `>=16` while the package release has a narrower `16 || 18`
> - The `package.json` declares its `engines.node` as `16` while the package release has a narrower `^16.10.0`

Author: ```suggestion
Here are some examples:

| project                  | dependency   | result   |
|--------------------------|--------------|----------|
| `18`                     | `16 \|\| 18` | allowed  |
| `^18.10.0`               | `>=18`       | allowed  |
| `^16.10.0 \|\| >=18.0.0` | `>= 16.0.0`  | allowed  |
| `>=16`                   | `16 \|\| 18` | filtered |
| `16`                     | `^16.10.0`   | filtered |
```

> When using with `npm`, it's recommended:

Reviewer: ```suggestion
When using with `npm`, we recommend you:
```
